specs/AppNotifications/AppNotificationContentSpec/AppNotificationBuilder-spec.md

```diff
@@ -0,0 +1,1170 @@
+# App Notification Content Builder
+
+# Background
+
+App Notifications are UI popups that contain rich text, controls and images to display a message to
+the user. With the Windows App SDK, developers can define an XML payload and display an
+AppNotification using the AppNotificationManager::Show API, but defining these XML payloads can be
+tedious when written by hand.
+
+To generate an AppNotification like this one:
+
+![App Notification With Avatar Image](AppNotificationWithAvatarImage.png)
+
+Developers have to write something like this:
+
+```c++
+int sequenceId {1234};
+std::wstring pathToImage {LR"(path\to\my\image.png)"};
+
+winrt::hstring xmlPayload{
+    L"<toast launch = \"action=AppNotificationClick;sequence=" + to_wstring(sequenceId) + L"\">\
+        <visual>\
+            <binding template = \"ToastGeneric\">\
+                <image placement = \"appLogoOverride\" hint-crop=\"circle\" src = \"" + pathToImage + L"\"/>\
+                <text>App Notification with Avatar Image</text>\
+                <text>This is an example message using XML</text>\
+            </binding>\
+        </visual>\
+        <actions>\
+            <action\
+                content = \"Open App\"\
+                arguments = \"action=OpenApp&amp;sequence=" + to_wstring(sequenceId) + L"\"/>\
+        </actions>\
+    </toast>" };
+
+auto appNotification{ winrt::AppNotification(xmlPayload) };
+winrt::AppNotificationManager::Default().Show(appNotication);
+```
+
+# Description
+
+To improve the AppNotification construction experience, the Windows App SDK is introducing
+AppNotificationContent, a builder that constructs simple or rich App Notifications for the
+developer.
+
+```c++
+AppNotificationContent()
+    .AddArgument(L"AppNotificationClick")
+    .AddArgument(L"sequence", 1234)
+    .SetAppLogoOverride(winrt::Windows::Foundation::Uri(L"http://www.contoso.com"), ImageCrop::Circle)
+    .AddText(L"App Notification with Avatar Image")
+    .AddText(L"This is an example message")
+    .AddButton(AppNotificationButton(L"Open App")
+        .AddArgument(L"action", "OpenAppButton")
+        .AddArgument(L"sequence", sequenceId));
+```
+
+# API Components
+
+# AppNotificationContent
+
+The AppNotificationContent component sets up the xml for a \<toast\>. This is the wrapper content
+that will include any component adding to the AppNotification UI.
+
+**WinAppSDK 1.2 \<toast\> Schema**:
+
+```c#
+<toast
+    launch? = string
+    duration? = "long" | "short"
+    displayTimeStamp? = string
+    scenario? = "reminder" | "alarm" | "incomingCall" | "urgent"
+    useButtonStyle? = boolean>
+      <!-- Child elements -->
+    visual, audio?, actions?
+</toast>
+```
+
+These attributes are abstracted away through the AppNotificationContent component APIs:
+
+-   AppNotificationContent();
```

Comment:
I'm not sure if you mention this anywhere. But the implication of using Builder Syntax is that ordering doesn't matter right for the API calls? Would be great to confirm that in the doc somewhere.

Reply:
It seems to me that order does matter. For instance in the AddText examples, it's clear the first AddText appear first, and is implicitly set to be bold.

Reply:
AddText orders them as they are added. I will make this explicit in the spec. This is the same process used for buttons and input (textbox/comboBox) components so I'll add it there too.